File: cascadeflow/providers/ollama.py
"""Ollama provider implementation for local LLM serving."""

import os
import time
from typing import Optional, Dict, Any

import httpx

from .base import BaseProvider, ModelResponse
from ..exceptions import ProviderError, ModelError


class OllamaProvider(BaseProvider):
    """
    Ollama provider for local LLM serving.

    Supports: Llama 3, Mistral, CodeLlama, Phi-3, Gemma, and 100+ other models.

    Benefits:
    - 100% FREE (no API costs)
    - Unlimited requests (no rate limits)
    - Privacy (runs locally)
    - Fast (no network latency)
    - Works with speculative cascading (automatic logprobs estimation)

    Requirements:
    - Ollama installed: https://ollama.com/download
    - Model pulled: `ollama pull llama3:8b`

    Example:
        >>> provider = OllamaProvider()
        >>> response = await provider.complete(
        ...     prompt="What is AI?",
        ...     model="llama3:8b"
        ... )
        >>> print(response.content)
        >>> print(f"Cost: ${response.cost}")  # Always $0.00!

        >>> # With logprobs (automatically estimated)
        >>> response = await provider.complete(
        ...     prompt="What is AI?",
        ...     model="llama3:8b",
        ...     logprobs=True  # Works via automatic fallback!
        ... )
        >>> print(response.logprobs)  # Estimated from temperature
    """

    def __init__(
            self,
            api_key: Optional[str] = None,
            base_url: Optional[str] = None
    ):
        """
        Initialize Ollama provider.

        Args:
            api_key: Not needed for Ollama (kept for interface compatibility)
            base_url: Ollama server URL. Defaults to http://localhost:11434
        """
        super().__init__(api_key)
        self.base_url = base_url or os.getenv("OLLAMA_HOST", "http://localhost:11434")

        # Initialize HTTP client (no auth needed for Ollama)
        self.client = httpx.AsyncClient(
            timeout=120.0  # Longer timeout for local inference
        )

    def _check_logprobs_support(self) -> bool:
        """Ollama does NOT support logprobs - uses automatic fallback."""
        return False

    async def complete(
            self,
            prompt: str,
            model: str,
            max_tokens: int = 4096,
            temperature: float = 0.7,
            system_prompt: Optional[str] = None,
            **kwargs
    ) -> ModelResponse:
        """
        Complete a prompt using Ollama.

        Args:
            prompt: User prompt
            model: Model name (e.g., 'llama3:8b', 'mistral:7b', 'codellama:7b')
            max_tokens: Maximum tokens to generate
            temperature: Sampling temperature (0-2)
            system_prompt: Optional system prompt
            **kwargs: Additional Ollama parameters
                     NEW: logprobs (bool) - Enable estimated logprobs (automatic fallback)

        Returns:
            ModelResponse with standardized format (with estimated logprobs if requested)

        Raises:
            ProviderError: If Ollama server not reachable
            ModelError: If model not found or execution fails
        """
        start_time = time.time()

        # Extract logprobs parameter (for automatic fallback)
        logprobs_enabled = kwargs.pop('logprobs', False)
        kwargs.pop('top_logprobs', None)  # Not used, but remove if present

        # Build request payload (Ollama format)
        payload = {
            "model": model,
            "prompt": prompt,
            "stream": False,  # We'll add streaming later
            "options": {
                "temperature": temperature,
                "num_predict": max_tokens,  # Ollama's name for max_tokens
            }
        }

        # Add system prompt if provided
        if system_prompt:
            payload["system"] = system_prompt

        # Add any additional options
        if kwargs:
            payload["options"].update(kwargs)

        try:
            # Make API request to Ollama
            response = await self.client.post(
                f"{self.base_url}/api/generate",
                json=payload
            )
            response.raise_for_status()

            data = response.json()

            # Extract response
            content = data.get("response", "")

            # Ollama doesn't return token counts, estimate them
            prompt_tokens = len(prompt) // 4
            completion_tokens = len(content) // 4
            tokens_used = prompt_tokens + completion_tokens

            # Calculate latency
            latency_ms = (time.time() - start_time) * 1000

            # Calculate cost (always $0 for Ollama!)
            cost = 0.0

            # Calculate confidence
            confidence = self.calculate_confidence(content, data)

            # Build base response
            model_response = ModelResponse(
                content=content,
                model=model,
                provider="ollama",
                cost=cost,  # FREE!
                tokens_used=tokens_used,
                confidence=confidence,
                latency_ms=latency_ms,
                metadata={
                    "done": data.get("done"),
                    "total_duration": data.get("total_duration"),
                    "load_duration": data.get("load_duration"),
                    "prompt_eval_count": data.get("prompt_eval_count"),
                    "eval_count": data.get("eval_count"),
                }
            )

            # Add estimated logprobs if requested
            # Ollama doesn't support real logprobs, so we ALWAYS use fallback
            if logprobs_enabled:
                model_response = self.add_logprobs_fallback(
                    model_response,
                    temperature,
                    base_confidence=0.75  # Local models slightly lower than API models
                )

            return model_response

        except httpx.HTTPStatusError as e:
            if e.response.status_code == 404:
                raise ModelError(
                    f"Model '{model}' not found. Did you run 'ollama pull {model}'?",
                    model=model,
                    provider="ollama"
                )
            else:
                raise ProviderError(
                    f"Ollama API error: {e.response.status_code}",
                    provider="ollama",
                    original_error=e
                )
        except httpx.ConnectError as e:
            raise ProviderError(
                "Cannot connect to Ollama. Is it running? Try: 'ollama serve'",
                provider="ollama",
                original_error=e
            )
        except httpx.RequestError as e:
            raise ProviderError(
                "Failed to connect to Ollama server",
                provider="ollama",
                original_error=e
            )
        except (KeyError, IndexError) as e:
            raise ModelError(
                f"Failed to parse Ollama response: {e}",
                model=model,
                provider="ollama"
            )

    def estimate_cost(self, tokens: int, model: str) -> float:
        """
        Estimate cost for Ollama model.

        Ollama is always FREE!

        Args:
            tokens: Total tokens (not used)
            model: Model name (not used)

        Returns:
            Always 0.0 (FREE!)
        """
        return 0.0  # Ollama is free!

    def calculate_confidence(
            self,
            response: str,
            metadata: Optional[Dict[str, Any]] = None
    ) -> float:
        """
        Calculate confidence score for Ollama response.

        Uses response length and completion status as heuristics.

        Args:
            response: Model response text
            metadata: Response metadata from Ollama

        Returns:
            Confidence score (0-1)
        """
        # Start with base confidence
        confidence = super().calculate_confidence(response)

        if metadata:
            # Check if generation completed successfully
            if metadata.get("done") == True:
                confidence = min(1.0, confidence + 0.05)

            # Ollama doesn't provide confidence scores,
            # so we rely on response quality heuristics

        return confidence

    async def list_models(self) -> list[str]:
        """
        List available models in Ollama.

        Returns:
            List of model names
        """
        try:
            response = await self.client.get(f"{self.base_url}/api/tags")
            response.raise_for_status()
            data = response.json()
            return [model["name"] for model in data.get("models", [])]
        except Exception as e:
            raise ProviderError(
                f"Failed to list Ollama models: {e}",
                provider="ollama"
            )

    async def pull_model(self, model: str) -> None:
        """
        Pull a model from Ollama library.

        Args:
            model: Model name to pull (e.g., 'llama3:8b')
        """
        try:
            response = await self.client.post(
                f"{self.base_url}/api/pull",
                json={"name": model}
            )
            response.raise_for_status()
        except Exception as e:
            raise ProviderError(
                f"Failed to pull model '{model}': {e}",
                provider="ollama"
            )

    async def __aenter__(self):
        """Async context manager entry."""
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Async context manager exit."""
        await self.client.aclose()

File: packages/core/src/providers/ollama.ts
/**
 * Ollama provider implementation
 *
 * Supports: Llama 3, Llama 3.1, Llama 3.2, Mistral, CodeLlama, Phi-3, Gemma, Qwen, DeepSeek, and 100+ models
 * Pricing: FREE (local execution, no API costs)
 * Features: Local model serving, privacy-first, tool calling support
 *
 * Ollama is perfect for development, testing, and privacy-sensitive applications
 */

import { BaseProvider, type ProviderRequest } from './base';
import type { ProviderResponse, Tool, Message, ReasoningModelInfo } from '../types';
import type { ModelConfig } from '../config';

/**
 * Detect if model is DeepSeek-R1 reasoning model
 *
 * @param modelName - Model name to check
 * @returns Model capabilities
 */
export function getReasoningModelInfo(modelName: string): ReasoningModelInfo {
  const name = modelName.toLowerCase();

  // DeepSeek-R1 - Chain-of-thought reasoning model
  // Variations: deepseek-r1, deepseek-r1:latest, deepseek-r1:8b, deepseek-r1:32b, deepseek-r1:70b
  if (name.includes('deepseek-r1') || name.includes('deepseek_r1')) {
    return {
      isReasoning: true,
      provider: 'ollama',
      supportsStreaming: true,
      supportsTools: true,
      supportsSystemMessages: true,
      supportsExtendedThinking: false,
      requiresThinkingBudget: false,
    };
  }

  // Standard models (no reasoning)
  return {
    isReasoning: false,
    provider: 'ollama',
    supportsStreaming: true,
    supportsTools: true,
    supportsSystemMessages: true,
    supportsExtendedThinking: false,
    requiresThinkingBudget: false,
  };
}

/**
 * Ollama provider for local LLM serving
 */
export class OllamaProvider extends BaseProvider {
  readonly name = 'ollama';
  private baseUrl: string;
  private timeout: number;

  constructor(config: ModelConfig) {
    super(config);
    this.baseUrl = config.baseUrl || 'http://localhost:11434';
    this.timeout = 300000; // 5 minutes for local inference
  }

  async generate(request: ProviderRequest): Promise<ProviderResponse> {
    try {
      const messages = this.normalizeMessages(request.messages);
      const ollamaMessages = this.convertToOllamaMessages(messages, request.systemPrompt);
      const tools = request.tools ? this.convertTools(request.tools) : undefined;

      // Ollama uses /api/chat endpoint for tool calling
      const response = await fetch(`${this.baseUrl}/api/chat`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          model: request.model || this.config.name,
          messages: ollamaMessages,
          stream: false,
          options: {
            temperature: request.temperature ?? this.config.temperature ?? 0.7,
            num_predict: request.maxTokens || this.config.maxTokens || 1000,
          },
          tools,
          ...request.extra,
        }),
        signal: AbortSignal.timeout(this.timeout),
      });

      if (!response.ok) {
        const error = await response.json().catch(() => ({ error: response.statusText }));
        throw new Error(`Ollama API error (${response.status}): ${JSON.stringify(error)}`);
      }

      const data = await response.json() as any;
      const message = data.message || {};

      // Parse tool calls if present
      const toolCalls = message.tool_calls?.map((tc: any, idx: number) => ({
        id: `call_${idx}`, // Ollama doesn't provide IDs
        type: 'function' as const,
        function: {
          name: tc.function?.name || '',
          arguments: typeof tc.function?.arguments === 'string'
            ? tc.function.arguments
            : JSON.stringify(tc.function?.arguments || {}),
        },
      }));

      // Estimate tokens (Ollama doesn't return token counts)
      const content = message.content || '';
      const promptTokens = ollamaMessages.reduce(
        (sum, msg) => sum + (msg.content?.length || 0),
        0
      ) / 4;
      const completionTokens = content.length / 4;

      return {
        content,
        model: request.model || this.config.name,
        usage: {
          prompt_tokens: Math.ceil(promptTokens),
          completion_tokens: Math.ceil(completionTokens),
          total_tokens: Math.ceil(promptTokens + completionTokens),
        },
        finish_reason: data.done ? 'stop' : 'incomplete',
        tool_calls: toolCalls,
        raw: data,
      };
    } catch (error) {
      throw this.formatError(error);
    }
  }

  calculateCost(_promptTokens: number, _completionTokens: number, _model: string): number {
    // Ollama is always free (local execution)
    return 0.0;
  }

  /**
   * Convert generic messages to Ollama format
   */
  private convertToOllamaMessages(messages: Message[], systemPrompt?: string): any[] {
    const ollamaMessages: any[] = [];

    if (systemPrompt) {
      ollamaMessages.push({ role: 'system', content: systemPrompt });
    }

    for (const msg of messages) {
      if (msg.role === 'system') {
        ollamaMessages.push({ role: 'system', content: msg.content });
      } else if (msg.role === 'user') {
        ollamaMessages.push({ role: 'user', content: msg.content });
      } else if (msg.role === 'assistant') {
        ollamaMessages.push({ role: 'assistant', content: msg.content });
      } else if (msg.role === 'tool') {
        // Ollama tool result format
        ollamaMessages.push({
          role: 'tool',
          content: msg.content,
        });
      }
    }

    return ollamaMessages;
  }

  /**
   * Convert generic tools to Ollama format (OpenAI-compatible)
   */
  private convertTools(tools: Tool[]): any[] {
    return tools.map((tool) => ({
      type: 'function',
      function: {
        name: tool.function.name,
        description: tool.function.description,
        parameters: tool.function.parameters,
      },
    }));
  }

  /**
   * Check if provider is available (Ollama doesn't require API key)
   */
  isAvailable(): boolean {
    return true; // Ollama doesn't require API key
  }

  /**
   * Override getApiKey to not require API key
   */
  protected getApiKey(): string {
    return ''; // Ollama doesn't need API key
  }

  /**
   * List available models from Ollama server
   */
  async listModels(): Promise<string[]> {
    try {
      const response = await fetch(`${this.baseUrl}/api/tags`);
      if (!response.ok) {
        throw new Error(`Failed to list models: ${response.statusText}`);
      }
      const data = await response.json() as any;
      return data.models?.map((m: any) => m.name) || [];
    } catch (error) {
      throw new Error(`Failed to list Ollama models: ${error}`);
    }
  }
}
